### ISSUE TITLE
Test on custom dataset such as DNA-Rendering

### ISSUE THREAD

**Issue by MyNiuuu** [wenj/LIFe-GoM#3]
Hi! Thank you for your amazing work! 

I would like to test your method on custom data, for example DNA-rendering. Besides the multi-view images and masks, what other preparations are necessary? For example, I noticed that in the Thuman2.0 dataset, a scene is structured as follows:

<img width="530" alt="Image" src="https://github.com/user-attachments/assets/bb354599-df49-44c0-9bff-56d5e3fb987e" />

Does this mean that I also need to prepare `cameras.pkl`, `canonical_joints.pkl`, and `mesh_infos.pkl` ?

Additionally, I saw that your paper includes results on DNA-rendering. Would it be possible for you to release the preprocessing code to generate the necessary `.pkl` files? I already have access to the original DNA-Rendering dataset.

Thank you very much for your help!

**Comment by wenj** [comment]
Thanks for your interest! The data structure follows what is produced [here](https://github.com/wenj/GoMAvatar/blob/main/scripts/prepare_zju-mocap/prepare_dataset.py). You'll need to modify SMPL to SMPL-X parameterizations.

**Comment by MyNiuuu** [comment]
Hi! Thanks for your quick reply!

Based on your provided script ([[prepare_dataset.py](https://github.com/wenj/GoMAvatar/blob/main/scripts/prepare_zju-mocap/prepare_dataset.py)](https://github.com/wenj/GoMAvatar/blob/main/scripts/prepare_zju-mocap/prepare_dataset.py)), I wrote my own preprocessing code for DNA-Rendering to generate the necessary `.pkl` files and then ran the inference script on DNA-Rendering. However, the results I obtained were as follows:

<img width="734" alt="Image" src="https://github.com/user-attachments/assets/0cf225ba-2d40-472d-a7da-a486121a0212" />

I suspect that the issue might be related to errors in my preprocessing scripts when generating the `.pkl` files.

Currently I plan to build upon your method for my work on DNA-Rendering. Since I see that you've conducted experiments on DNA-Rendering in your paper, I was wondering if it would be possible for you to directly release the preprocessing codes for generating the necessary `.pkl` files for DNA-Rendering so that I could use it to run your method on the dataset? That would be incredibly helpful, and I would really appreciate it!

**Comment by wenj** [comment]
Here's the script I used to preprocess the data:
```
import os
import sys
sys.path.append('../../') # append the path of LIFe-GoM/

from shutil import copyfile

import pickle
import yaml
import numpy as np
from tqdm import tqdm
import trimesh
from pathlib import Path
from smplx import SMPLX
import torch
from PIL import Image
import cv2
import seaborn as sns
import json

from utils.smpl.smpl_numpy import SMPL
from utils.file_util import split_path
from utils.image_util import load_image, save_image, to_3ch_image
from utils.body_util import _rvec_to_rmtx, body_pose_to_body_RTs, body_RTs_to_global_RTs, SMPLX_PARENT

from absl import app
from absl import flags

from smc_reader import SMCReader
from smc_annots_reader import SMCReader as SMCAnnotsReader

FLAGS = flags.FLAGS

flags.DEFINE_string('cfg',
                    'config.yaml',
                    'the path of config file')

flags.DEFINE_string('subject_id',
                    '0008_01',
                    'the path of config file')

flags.DEFINE_integer('frame_id',
                    0,
                    'the path of config file')

MODEL_DIR = '../../utils/smpl/models'

debug = True


def parse_config():
    config = None
    with open(FLAGS.cfg, 'r') as file:
        config = yaml.full_load(file)

    return config


def prepare_dir(output_path, name):
    out_dir = os.path.join(output_path, name)
    os.makedirs(out_dir, exist_ok=True)

    return out_dir


def get_mask(subject_dir, img_name):
    msk_path = os.path.join(subject_dir,
                            img_name)[:-4] + '.png'
    msk = np.array(load_image(msk_path))[:, :, 0]
    msk = (msk > 0.95).astype(np.uint8)
    msk[msk == 1] = 255

    return msk


def main(argv):
    del argv  # Unused.

    cfg = parse_config()
    dataset_dir = cfg['dataset']['dna_rendering_path']

    subject = FLAGS.subject_id
    frame_id = FLAGS.frame_id
    sex = 'neutral'

    rd = SMCReader(os.path.join(dataset_dir, 'dna_rendering_part1_main', subject + '.smc'))
    annots_rd = SMCAnnotsReader(os.path.join(dataset_dir, 'dna_rendering_part1_annotations', subject + '.smc'))

    # prepare output directories
    output_path = os.path.join(cfg['output']['dir'], f'{subject}_{frame_id:06d}')
    out_img_dir = prepare_dir(output_path, 'images')
    out_mask_dir = prepare_dir(output_path, 'masks')
    if debug:
        out_debug_dir = prepare_dir(output_path, 'debug')

    # copy config file
    copyfile(FLAGS.cfg, os.path.join(output_path, 'config.yaml'))

    params = annots_rd.get_SMPLx(Frame_id=frame_id)

    smplx_model = SMPLX(model_path=os.path.join(MODEL_DIR, 'SMPLX_{}.npz'.format(sex.upper())), use_pca=False)
    parents = {}
    for i, parent_i in enumerate(smplx_model.parents.detach().cpu().numpy()):
        parents[i] = parent_i

    cameras = {}
    mesh_infos = {}

    with torch.no_grad():
        output = smplx_model(
            betas=torch.tensor(params['betas']).to(torch.float32)[None],
            return_full_pose=True,
        )
        tpose_joints = output.joints.detach().cpu().numpy()[0, :55]
        tpose = output.full_pose.detach()
        tpose_vertices = output.vertices.detach().cpu().numpy()[0]

    # move pelvis root rotation and transl to Rh and Th
    Rh = params['fullpose'][0]
    Th = params['transl'] + tpose_joints[0] - _rvec_to_rmtx(Rh) @ tpose_joints[0]

    with torch.no_grad():
        params['fullpose'] = params['fullpose'] + smplx_model.pose_mean.detach().cpu().numpy().reshape(55, 3)
        output = smplx_model(
            transl=torch.tensor(params['transl']).to(torch.float32)[None],
            global_orient=torch.tensor(params['fullpose'][0]).to(torch.float32)[None],
            body_pose=torch.tensor(params['fullpose'][1:22]).to(torch.float32)[None],
            betas=torch.tensor(params['betas']).to(torch.float32)[None],
            jaw_pose=torch.tensor(params['fullpose'][22]).to(torch.float32)[None],
            leye_pose=torch.tensor(params['fullpose'][23]).to(torch.float32)[None],
            reye_pose=torch.tensor(params['fullpose'][24]).to(torch.float32)[None],
            left_hand_pose=torch.tensor(params['fullpose'][25:40]).to(torch.float32)[None],
            right_hand_pose=torch.tensor(params['fullpose'][40:55]).to(torch.float32)[None],
            return_full_pose=True,
        )
    joints = output.joints.detach().cpu().numpy()[0, :55]
    full_pose = (output.full_pose - smplx_model.pose_mean).detach().cpu().numpy().reshape(55, 3)
    full_pose[0] = 0.

    for camera_id in range(48):
        out_name = 'frame_{:06d}'.format(camera_id)
        img = rd.get_img('Camera_5mp', camera_id, 'color', Frame_id=frame_id)[:, :, [2, 1, 0]]
        mask = annots_rd.get_mask(camera_id, Frame_id=frame_id)

        camera = annots_rd.get_Calibration(camera_id)
        E = np.linalg.inv(camera['RT'])
        K = camera['K']
        D = camera['D']

        # write image and mask
        save_image(mask[:, :, None].repeat(3, axis=-1),
                   os.path.join(out_mask_dir, out_name + '.png'))
        out_image_path = os.path.join(out_img_dir, '{}.png'.format(out_name))
        save_image(img, out_image_path)

        cameras[out_name] = {
            'intrinsics': K,
            'extrinsics': E,
            'distortions': D,
        }

        mesh_infos[out_name] = {
            'Rh': Rh,
            'Th': Th,
            'poses': full_pose,
            'joints': joints,
            'tpose_joints': tpose_joints
        }

        if debug:
            print(out_name)
            img = np.ascontiguousarray(img, dtype=np.uint8)
            img = cv2.undistort(img, K, D)
            Rs, Ts = body_pose_to_body_RTs(mesh_infos[out_name]['poses'], tpose_joints, use_smplx=True)
            Rs, Ts = body_RTs_to_global_RTs(Rs, Ts, use_smplx=True)
            # print(Ts @ cv2.Rodrigues(Rh)[0].T + Th[None, :])
            # print(Ts, joints)  # should be the same

            joints_world = cv2.Rodrigues(Rh)[0] @ Ts.T + Th[:, None]
            joints_world = joints_world.T
            # for j1, j2 in zip(joints_final, joints_world):
            #     print(j1, j2)
            # joints_world = joints
            joints_world = np.concatenate([joints_world, np.ones_like(joints_world[:, :1])], axis=-1)
            joints_cam = E @ joints_world.T
            joints_cam = joints_cam.T
            joints_cam = joints_cam[:, :3] / joints_cam[:, 3:]
            joints_2d = K @ joints_cam.T
            joints_2d = joints_2d.T
            joints_2d = joints_2d[:, :2] / joints_2d[:, 2:]
            # print(joints_2d)

            colors = np.array(sns.color_palette("tab10", joints_2d.shape[0])) * 255
            for i, (pt, color) in enumerate(zip(joints_2d, colors)):
                cv2.circle(img, (int(pt[0]), int(pt[1])), radius=1,
                           color=(int(color[0]), int(color[1]), int(color[2])), thickness=-1)
                if i == 0:
                    continue
                parent = SMPLX_PARENT[i]
                cv2.line(img, (int(joints_2d[parent][0]), int(joints_2d[parent][1])), (int(pt[0]), int(pt[1])),
                         (255, 0, 0), 5)
            out_image_path = os.path.join(out_debug_dir, '{}.png'.format(out_name))
            save_image(img, out_image_path)

    # write camera infos
    with open(os.path.join(output_path, 'cameras.pkl'), 'wb') as f:
        pickle.dump(cameras, f)

    # write mesh infos
    with open(os.path.join(output_path, 'mesh_infos.pkl'), 'wb') as f:
        pickle.dump(mesh_infos, f)

    # write canonical joints
    template_joints = tpose_joints
    vertex = tpose_vertices
    weights = smplx_model.lbs_weights.detach().cpu().numpy()

    mesh = trimesh.Trimesh(vertex, faces=smplx_model.faces)
    edges = mesh.edges
    with open(os.path.join(output_path, 'canonical_joints.pkl'), 'wb') as f:
        pickle.dump(
            {
                'vertex': vertex,
                'joints': template_joints,
                'weights': weights,
                'edges': edges,
                'faces': smplx_model.faces,
            }, f)


if __name__ == '__main__':
    app.run(main)

```

The config file `config.yaml` looks like this:
```
dataset:
  dna_rendering_path: YOUR_PATH_TO_DNA_RENDERING/Part\ 1
  sex: 'male'
  input_frame: 0

output:
  dir: YOUR_PATH_TO_SAVE_PROCESSED_DATA

```

Run `python prepare_dataset.py --subject_id $SUBJECT_ID --frame_id $FRAME_ID`, e.g., `python prepare_dataset.py --subject_id 0008_01 --frame_id 0`. You may also need to download [SMPL-X](https://smpl-x.is.tue.mpg.de/). The imported smc_reader and  smc_annots_reader are from DNA-Rendering.

You could set `debug=True` to visualize the alignment of SMPL-X parameters and images.

The script is not cleaned up yet. Please let me know if you have further questions. Hope this helps!

**Comment by MyNiuuu** [comment]
Hi! Thank you so much for the scripts, now I can get the right results on DNA-Rendering, so many thanks for ur help!
